src/main/scala/sbtprotoc/ProtocPlugin.scala

```diff
@@ -522,36 +521,102 @@ object ProtocPlugin extends AutoPlugin {
+                        (newFilesInfo, sandboxedClassLoader(resolved))
+                      }
+                    }
+                  }
+                )
+            classloader
         }.value
 
+      val targets = (key / PB.targets).value
+      val arguments = Arguments(
+        PB.protocVersion.value,
+        (key / PB.includePaths).value,
+        protocOptions = (key / PB.protocOptions).value ++ nativePluginsArgs,
+        deleteTargetDirectory = (key / PB.deleteTargetDirectory).value,
+        targets = targets.map { target =>
+          (
+            target.generator.name,
+            target.generator.suggestedDependencies,
+            target.outputPath,
+            target.options
+          )
+        }
+      )
+
       def compileProto(): Set[File] =
         compile(
           (key / PB.runProtoc).value,
           schemas,
-          (key / PB.includePaths).value,
-          (key / PB.protocOptions).value ++ nativePluginsArgs,
-          (key / PB.targets).value,
-          (key / PB.deleteTargetDirectory).value,
-          (key / streams).value.log,
+          arguments.includePaths,
+          arguments.protocOptions,
+          targets,
+          arguments.deleteTargetDirectory,
+          log,
           classLoader
         )
 
-      val cachedCompile = FileFunction.cached(
-        cacheFile,
-        inStyle = FilesInfo.lastModified,
-        outStyle = FilesInfo.exists
-      ) { (in: Set[File]) => compileProto() }
+      import CacheImplicits._
+      val cachedCompile = Tracked.inputChanged[(Arguments, FilesInfo[ModifiedFileInfo]), Set[File]](
+        cacheFile / "input"
+      ) { case (inChanged, _) =>
+        Tracked.diffOutputs(
+          cacheFile / "output",
+          FileInfo.exists
+        ) { outDiff: ChangeReport[File] =>
+          if (inChanged || outDiff.modified.nonEmpty) {
+            log.debug {
+              val inputInvalidation =
+                if (inChanged) Seq("input changed")
+                else Seq()
+              val outputInvalidation =
+                if (outDiff.modified.nonEmpty) Seq(s"output files changed ${outDiff.modified}")
+                else Seq()
+              s"Invalidating cache (${(inputInvalidation ++ outputInvalidation).mkString(" and ")})"
+            }
+            compileProto()
+          } else outDiff.checked
+        }
+      }
 
+      val resolver = (key / PB.artifactResolver).value
       if (PB.recompile.value) {
+        log.debug("Ignoring cache (PB.recompile := true)")
         compileProto().toSeq
       } else {
-        cachedCompile(schemas).toSeq
+        val sandboxedArtifactsClasspath =
+          targets
+            .map(_.generator)
+            .collect { case SandboxedJvmGenerator(_, artifact, _, _) => artifact }
+            .flatMap(resolver(_).allPaths.get) // artifact paths can be JARs or directories
+        val input = schemas ++ arguments.includePaths.allPaths.get ++ sandboxedArtifactsClasspath
```

Comment:
@thesamet FTR, as discovered through https://github.com/akka/akka-grpc/pull/1264/commits/9fb3944dbfafd9f099268a51d6e916068b10171f, this slightly changes the behavior as it evaluates the resolver(s) eagerly, even if they are no proto sources. I guess the previous behavior could be brought back by checking `schemas` emptiness earlier, but I am not sure it's worth it.